src/targets/gpu/quant_gemm.cpp

```diff
@@ -0,0 +1,187 @@
+#include <migraphx/gpu/quant_gemm.hpp>
+#include <migraphx/gpu/device/pack.hpp>
+#include <migraphx/gpu/context.hpp>
+#include <migraphx/generate.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace gpu {
+
+template <class... Ts>
+rocblas_status generic_rocblas_gemm_ex(Ts&&... xs)
+{
+    return rocblas_gemm_ex(std::forward<Ts>(xs)...);
+}
+
+template <class... Ts>
+rocblas_status generic_rocblas_batched_gemm_ex(Ts&&... xs)
+{
+    return rocblas_gemm_strided_batched_ex(std::forward<Ts>(xs)...);
+}
+
+template <class T>
+struct compute_rocblas_type
+{
+    using type = T;
+};
+
+template <class T>
+struct compute_rocblas_type<const T>
+{
+    using type = const typename compute_rocblas_type<T>::type;
+};
+
+template <>
+struct compute_rocblas_type<half>
+{
+    using type = rocblas_half;
+};
+
+template <class T>
+using rb_type = typename compute_rocblas_type<T>::type;
+
+template <class T>
+rb_type<T> to_rocblas_type(T x)
+{
+    return reinterpret_cast<const rb_type<T>&>(x);
+}
+
+template <class T>
+rb_type<T>* to_rocblas_type(T* x)
+{
+    return reinterpret_cast<rb_type<T>*>(x);
+}
+
+shape miopen_quant_gemm::compute_shape(const std::vector<shape>& inputs) const
+{
+    std::vector<shape> in_shapes(inputs);
+    in_shapes.erase(in_shapes.begin() + in_shapes.size() - 3, in_shapes.end());
+    check_shapes{in_shapes}.not_broadcasted();
+
+    return op.compute_shape(in_shapes);
+}
+
+argument miopen_quant_gemm::compute(context& ctx,
+                                    const shape& output_shape,
+                                    const std::vector<argument>& args) const
+{
+    bool transa     = args[0].get_shape().transposed();
+    bool transb     = args[1].get_shape().transposed();
+    auto n_dim      = output_shape.lens().size();
+    auto dim_1      = n_dim - 1;
+    auto dim_0      = n_dim - 2;
+    auto arg_num    = args.size();
+    rocblas_int lda = args[0].get_shape().strides()[transa ? dim_1 : dim_0];
+    rocblas_int ldb = args[1].get_shape().strides()[transb ? dim_1 : dim_0];
+    rocblas_int ldc = args[arg_num - 1].get_shape().strides()[dim_0];
+
+    if(!transb)
+    {
+        device::pack_a(ctx.get_stream().get(), args[arg_num - 2], args[1]);
+    }
+
+    // need to pack A in this scenario, use the algorithm to pack B in the
+    // comment of the API
+    if(transa)
+    {
+        device::pack_b(ctx.get_stream().get(), args[arg_num - 3], args[0]);
```

Comment:
This should be done as a separate operator and inserted during lowering.

Reply:
fixed